### PR TITLE
 [Threading] Implement multithreading at the host level

### PR DESCRIPTION
**Pull request by octaviansima**
This PR contains the necessary refactoring in order for `enclave_modelaggregator` to run in parallel. It defines new ecalls to load and return the model in enclave memory as well as to set the number of threads.

### CODE REVIEW

**Comment by chester-leung** [comment]
Thanks! Can you clean up/modify the PR such that only the changes necessary for multithreading are included? i.e. no need for the `docker` directory, the extra empty files with changed permissions, etc

**Review by podcastinator — CHANGES_REQUESTED** [review]
Thanks! Left a couple comments. Also, a bunch of files have weird permissions. Please revert them to 644?

File: server/enclave/ecalls.cpp [hunk 1]
```diff
@@ -142,26 +158,28 @@ void enclave_modelaggregator(uint8_t*** encrypted_accumulator,
...
     *new_params_length = serialized_buffer_size;
     size_t item_lengths[3] = {*new_params_length, CIPHER_IV_SIZE, CIPHER_TAG_SIZE};
-    for (int i = 0; i < encryption_metadata_length; i++) {
+    for (int i = 0; i < ENCRYPTION_METADATA_LENGTH; i++) {
         (*encrypted_new_params_ptr)[i] = (uint8_t*) oe_host_malloc(item_lengths[i] * sizeof(uint8_t));
         memcpy((void *) (*encrypted_new_params_ptr)[i], (const void*) encrypted_new_params[i], item_lengths[i] * sizeof(uint8_t));
     }
 
-    delete_double_ptr(encrypted_new_params, encryption_metadata_length);
+    delete_double_ptr(encrypted_new_params, ENCRYPTION_METADATA_LENGTH);
```
Comment (chester-leung) [inline server/enclave/ecalls.cpp:184]: I believe you need to reset the globals after you finish everything else?

**Review by chester-leung — COMMENTED** [review]
Can you please also revert the permissions of `README.md`, `client/README.md`, `common/encryption/encrypt.h`, and `common/utils.h` to 644?

File: server/enclave/ecalls.cpp [hunk 2]
```diff
@@ -26,6 +26,18 @@ if (!oe_is_outside_enclave((ptr), size)) {                \
...
 }
 
+// Defined in modelaggregator.conf
+static const int MAX_TCS = 32;
+
+// Ciphertext, IV, and tag are required for decryption
+static const size_t ENCRYPTION_METADATA_LENGTH = 3;
+
+// Global variables stored for threading
+static vector<map<string, vector<double>>> g_accumulator;
```
Comment (chester-leung) [inline server/enclave/ecalls.cpp:36]: Maybe you can also rename these global variables to `global_*` instead of `g_*` for clarity?

Comment (octaviansima) [reply server/enclave/ecalls.cpp:36]: Using `g_*` was a popular way to do it in many style guides I saw, such as: https://users.ece.cmu.edu/~eno/coding/CppCodingStandard.html#gnames

Comment (chester-leung) [reply server/enclave/ecalls.cpp:36]: Oh perfect, this is great then!

**Review by podcastinator — APPROVED** [review]

**Pull request merged**